### PR TITLE
feat(connector): [AUTHORIZEDOTNET] Implement non-zero mandates

### PR DESCRIPTION
**Pull request by deepanshu-iiitu**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Non-zero dollar mandates are implemented for connector Authorizedotnet.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

https://github.com/juspay/hyperswitch/issues/4703

## How did you test it?

- Payment Intent Create (Non-Zero Mandate):

Request:
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: API_KEY_HERE' \
--data-raw '{
    "amount": 370,
    "currency": "USD",
    "confirm": false,
    "customer_id": "cust6",
    "email": "something2@gmail.com",
    "name": "Joseph Doe",
    "description": "Its my first payment request",
    "setup_future_usage":"off_session"
}'
```

Response:
```
 {
    "payment_id": "pay_lx2DcRsYllb2QenBQfze",
    "merchant_id": "merchant_1716370796",
    "status": "requires_payment_method",
    "amount": 370,
    "net_amount": 370,
    "amount_capturable": 0,
    "amount_received": null,
    "connector": null,
    "client_secret": "pay_lx2DcRsYllb2QenBQfze_secret_X1riA15AgjzqEahbR8mv",
    "created": "2024-05-27T11:07:15.600Z",
    "currency": "USD",
    "customer_id": "cust6",
    "customer": {
        "id": "cust6",
        "name": "Joseph Doe",
        "email": "something2@gmail.com",
        "phone": null,
        "phone_country_code": null
    },
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": "off_session",
    "off_session": null,
    "capture_on": null,
    "capture_method": null,
    "payment_method": null,
    "payment_method_data": null,
    "payment_token": null,
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": "something2@gmail.com",
    "name": "Joseph Doe",
    "phone": null,
    "return_url": null,
    "authentication_type": null,
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": null,
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "cust6",
        "created_at": 1716808035,
        "expires": 1716811635,
        "secret": "epk_339c8ebb1d9f430796cf8273ef72f78f"
    },
    "manual_retry_allowed": null,
    "connector_transaction_id": null,
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": null,
    "payment_link": null,
    "profile_id": "pro_ZjmEyV32O7pOQkBlB510",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": null,
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-05-27T11:22:15.600Z",
    "fingerprint": null,
    "browser_info": null,
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-05-27T11:07:15.690Z",
    "charges": null,
    "frm_metadata": null
}
```

- Payment Confirm:

Request:
```
curl --location 'http://localhost:8080/payments/pay_lx2DcRsYllb2QenBQfze/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: API_KEY_HERE' \
--data '{
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4111111111111111",
            "card_exp_month": "07",
            "card_exp_year": "26",
            "card_holder_name": "Joseph Does",
            "card_cvc": "123"
        }
    },
    "mandate_data": {
        "customer_acceptance": {
            "acceptance_type": "offline",
            "accepted_at": "1963-05-03T04:07:52.723Z",
            "online": {
                "ip_address": "125.0.0.1",
                "user_agent": "amet irure esse"
            }
        },
        "mandate_type": {
            "multi_use": {
                "amount": 68607706,
                "currency": "USD",
                "start_date": "2023-04-21T00:00:00Z",
                "end_date": "2023-05-21T00:00:00Z",
                "metadata": {
                    "frequency": "13"
                }
            }
        }
    }
}'
```

Response:
```
{
    "payment_id": "pay_lx2DcRsYllb2QenBQfze",
    "merchant_id": "merchant_1716370796",
    "status": "succeeded",
    "amount": 370,
    "net_amount": 370,
    "amount_capturable": 0,
    "amount_received": 370,
    "connector": "authorizedotnet",
    "client_secret": "pay_lx2DcRsYllb2QenBQfze_secret_X1riA15AgjzqEahbR8mv",
    "created": "2024-05-27T11:07:15.600Z",
    "currency": "USD",
    "customer_id": "cust6",
    "customer": {
        "id": "cust6",
        "name": "Joseph Doe",
        "email": "something2@gmail.com",
        "phone": null,
        "phone_country_code": null
    },
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": "man_wxTMvSxaQPFC5uvg7Qmm",
    "mandate_data": {
        "update_mandate_id": null,
        "customer_acceptance": {
            "acceptance_type": "offline",
            "accepted_at": "1963-05-03T04:07:52.723Z",
            "online": {
                "ip_address": "125.0.0.1",
                "user_agent": "amet irure esse"
            }
        },
        "mandate_type": {
            "multi_use": {
                "amount": 68607706,
                "currency": "USD",
                "start_date": "2023-04-21T00:00:00.000Z",
                "end_date": "2023-05-21T00:00:00.000Z",
                "metadata": {
                    "frequency": "13"
                }
            }
        }
    },
    "setup_future_usage": "off_session",
    "off_session": null,
    "capture_on": null,
    "capture_method": null,
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "1111",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "411111",
            "card_extended_bin": null,
            "card_exp_month": "07",
            "card_exp_year": "26",
            "card_holder_name": "Joseph Does",
            "payment_checks": null,
            "authentication_data": null
        },
        "billing": null
    },
    "payment_token": null,
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": "something2@gmail.com",
    "name": "Joseph Doe",
    "phone": null,
    "return_url": null,
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": false,
    "connector_transaction_id": "80019080368",
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "80019080368",
    "payment_link": null,
    "profile_id": "pro_ZjmEyV32O7pOQkBlB510",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_6wtrG4tIah8tNezhNkT9",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-05-27T11:22:15.600Z",
    "fingerprint": null,
    "browser_info": {
        "language": null,
        "time_zone": null,
        "ip_address": "::1",
        "user_agent": null,
        "color_depth": null,
        "java_enabled": null,
        "screen_width": null,
        "accept_header": null,
        "screen_height": null,
        "java_script_enabled": null
    },
    "payment_method_id": "pm_ZN7bB0d33KWNWIUgVWod",
    "payment_method_status": null,
    "updated": "2024-05-27T11:07:32.165Z",
    "charges": null,
    "frm_metadata": null
}
```

- Payment Create (Using Mandate ID):

Request:
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: API_KEY_HERE' \
--data '{
    "amount": 141,
    "currency": "USD",
    "confirm": true,
    "customer_id": "cust6",
    "capture_method": "automatic",
    "mandate_id": "man_wxTMvSxaQPFC5uvg7Qmm",
    "off_session": true
}'
```

Response:
```
{
    "payment_id": "pay_8aI2C6oCZosMy1CQf2X9",
    "merchant_id": "merchant_1716370796",
    "status": "succeeded",
    "amount": 141,
    "net_amount": 141,
    "amount_capturable": 0,
    "amount_received": 141,
    "connector": "authorizedotnet",
    "client_secret": "pay_8aI2C6oCZosMy1CQf2X9_secret_gnro3S1Hlru3sUsAcdik",
    "created": "2024-05-27T11:07:53.638Z",
    "currency": "USD",
    "customer_id": "cust6",
    "customer": {
        "id": "cust6",
        "name": "Joseph Doe",
        "email": "something2@gmail.com",
        "phone": null,
        "phone_country_code": null
    },
    "description": null,
    "refunds": null,
    "disputes": null,
    "mandate_id": "man_wxTMvSxaQPFC5uvg7Qmm",
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "1111",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "411111",
            "card_extended_bin": null,
            "card_exp_month": "07",
            "card_exp_year": "26",
            "card_holder_name": "Joseph Does",
            "payment_checks": null,
            "authentication_data": null
        },
        "billing": null
    },
    "payment_token": "fa0a6a44-2e80-4953-971d-c70702e889d7",
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": "something2@gmail.com",
    "name": "Joseph Doe",
    "phone": null,
    "return_url": null,
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "cust6",
        "created_at": 1716808073,
        "expires": 1716811673,
        "secret": "epk_95141b60e629482ea657f64053afa3d2"
    },
    "manual_retry_allowed": false,
    "connector_transaction_id": "80019080384",
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "80019080384",
    "payment_link": null,
    "profile_id": "pro_ZjmEyV32O7pOQkBlB510",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_6wtrG4tIah8tNezhNkT9",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-05-27T11:22:53.638Z",
    "fingerprint": null,
    "browser_info": null,
    "payment_method_id": "pm_ZN7bB0d33KWNWIUgVWod",
    "payment_method_status": "active",
    "updated": "2024-05-27T11:07:54.663Z",
    "charges": null,
    "frm_metadata": null
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/connector/authorizedotnet/transformers.rs [hunk 1]
```diff
@@ -461,126 +482,168 @@ impl TryFrom<&AuthorizedotnetRouterData<&types::PaymentsAuthorizeRouterData>>
...
+                            (
+                                Some(PaymentDetails::CreditCard(CreditCardDetails {
+                                    card_number: (*ccard.card_number).clone(),
+                                    // expiration_date: format!("{expiry_year}-{expiry_month}").into(),
+                                    expiration_date: ccard.get_expiry_date_as_yyyymm("-"),
+                                    card_code: Some(ccard.card_cvc.clone()),
+                                })),
+                                Some(ProcessingOptions {
+                                    is_subsequent_auth: true,
+                                }),
```
Comment (SamraatBansal) [inline crates/router/src/connector/authorizedotnet/transformers.rs:610]: Should this be `true`?

**Review by SamraatBansal — CHANGES_REQUESTED** [review]
Refactor the code with clear `try_froms`

File: crates/router/src/connector/authorizedotnet/transformers.rs (hunk 1 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/authorizedotnet/transformers.rs]: Handled

File: crates/router/src/connector/authorizedotnet/transformers.rs [hunk 2]
```diff
@@ -510,16 +532,116 @@ impl TryFrom<&AuthorizedotnetRouterData<&types::PaymentsAuthorizeRouterData>>
...
+    fn try_from(
+        (item, network_trans_id): (
+            &AuthorizedotnetRouterData<&types::PaymentsAuthorizeRouterData>,
+            String,
+        ),
+    ) -> Result<Self, Self::Error> {
+        Ok(Self {
+            transaction_type: TransactionType::try_from(item.router_data.request.capture_method)?,
+            amount: item.amount,
+            currency_code: item.router_data.request.currency.to_string(),
```
Comment (srujanchikke) [inline crates/router/src/connector/authorizedotnet/transformers.rs:563]: Can we change this currency type to `common_enums::Currency` from string.

File: crates/router/src/connector/authorizedotnet/transformers.rs [hunk 3]
```diff
@@ -510,16 +532,116 @@ impl TryFrom<&AuthorizedotnetRouterData<&types::PaymentsAuthorizeRouterData>>
...
+            payment: Some(match item.router_data.request.payment_method_data {
+                domain::PaymentMethodData::Card(ref ccard) => {
+                    PaymentDetails::CreditCard(CreditCardDetails {
+                        card_number: (*ccard.card_number).clone(),
+                        expiration_date: ccard.get_expiry_date_as_yyyymm("-"),
+                        card_code: None,
```
Comment (srujanchikke) [inline crates/router/src/connector/authorizedotnet/transformers.rs:569]: Why is this card_code set to None ?

File: crates/router/src/connector/authorizedotnet/transformers.rs [hunk 4]
```diff
@@ -532,64 +654,87 @@ impl TryFrom<&AuthorizedotnetRouterData<&types::PaymentsAuthorizeRouterData>>
...
-        let bill_to = match profile {
-            Some(_) => None,
-            None => item
+        Ok(Self {
+            transaction_type: TransactionType::try_from(item.router_data.request.capture_method)?,
+            amount: item.amount,
+            currency_code: item.router_data.request.currency.to_string(),
+            payment: Some(PaymentDetails::CreditCard(CreditCardDetails {
+                card_number: (*ccard.card_number).clone(),
+                // expiration_date: format!("{expiry_year}-{expiry_month}").into(),
```
Comment (srujanchikke) [inline crates/router/src/connector/authorizedotnet/transformers.rs:728]: Can we remove this commented code ?

**Review by srujanchikke — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/authorizedotnet/transformers.rs (hunk 3 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/authorizedotnet/transformers.rs:569]: `card_code` is not required for payments which are done using `network_trans_id`

File: crates/router/src/connector/authorizedotnet/transformers.rs (hunk 4 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/authorizedotnet/transformers.rs:728]: Done

File: crates/router/src/connector/authorizedotnet/transformers.rs (hunk 2 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/authorizedotnet/transformers.rs]: Done

**Review by srujanchikke — APPROVED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Pull request merged**